### PR TITLE
Remove `git.Command.Run` and `git.Command.RunInDir*`

### LINKED ISSUE #18553: replace all git.Command.RunInPipeline... with RunWithContext
- [x]  Change git.cmd to RunWithContext (#18693)
        cmd.RunInDirTimeoutPipeline,
        cmd.RunInDirTimeoutEnvFullPipeline,
        cmd.RunInDirFullPipeline,
        ...
- [x] Delete cmd.RunInDir... func()

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follows
* https://github.com/go-gitea/gitea/pull/19266
* https://github.com/go-gitea/gitea/issues/18553

Close #18553

Now, only `RunWithContext` related functions.

This PR is a simple search & replace:
 * before: `stdout, err := RunInDir(path)`
 * now: `stdout, _, err := RunStdString(&git.RunOpts{Dir:path})`

### CODE REVIEW

File: integrations/git_test.go [hunk 1]
```diff
@@ -159,10 +159,11 @@ func lfsCommitAndPushTest(t *testing.T, dstPath string) (littleLFS, bigLFS strin
 			t.Skip()
 			return
 		}
+		var err error
```
Comment (6543) [inline integrations/git_test.go]:
```suggestion
```

File: integrations/git_test.go [hunk 2]
```diff
@@ -159,10 +159,11 @@ func lfsCommitAndPushTest(t *testing.T, dstPath string) (littleLFS, bigLFS strin
...
 		prefix := "lfs-data-file-"
-		_, err := git.NewCommand(git.DefaultContext, "lfs").AddArguments("install").RunInDir(dstPath)
+		_, _, err = git.NewCommand(git.DefaultContext, "lfs").AddArguments("install").RunWithContextString(&git.RunContext{Dir: dstPath})
```
Comment (6543) [inline integrations/git_test.go]:
```suggestion
		_, _, err := git.NewCommand(git.DefaultContext, "lfs").AddArguments("install").RunWithContextString(&git.RunContext{Dir: dstPath})
```

File: integrations/git_test.go [hunk 3]
```diff
@@ -671,7 +672,8 @@ func doCreateAgitFlowPull(dstPath string, ctx *APITestContext, baseBranch, headB
 		})
 
 		t.Run("Push", func(t *testing.T) {
-			_, err := git.NewCommand(git.DefaultContext, "push", "origin", "HEAD:refs/for/master", "-o", "topic="+headBranch).RunInDir(dstPath)
+			var err error
```
Comment (6543) [inline integrations/git_test.go]: same ...

File: integrations/git_test.go (hunk 1 above)
Comment (wxiaoguang) [reply integrations/git_test.go]: No, it doesn't work, the change is by purpose.

Because the return type of `RunWithXxx` is `RunError`, not `error`

Comment (wxiaoguang) [reply integrations/git_test.go]: More details: why using `RunError` but not `error`: it reminds the developers that such error contains the `Stderr`, it's different from a normal command error.


https://github.com/go-gitea/gitea/blob/f6145a69c4814d9aaaeac5d73c743b2f65bc586c/modules/git/command.go#L212-L241

Comment (wxiaoguang) [reply integrations/git_test.go]: I can remove some unnecessary `var err error`, just a minute.

**Review by 6543 — APPROVED** [review]

File: integrations/git_test.go (hunk 1 above)
Comment (wxiaoguang) [reply integrations/git_test.go]: All unnecessray `var err error` are removed. The remaining (about 7) are still necessary. I think it won't cause any misunderstanding now.

**Comment by wxiaoguang** [comment]
> I think it's better to have some rename.
> 
>     * rename `RunWithContext` -> `Run`
> 
>     * rename `RunWithContextString` -> `RunString`
> 
>     * rename `RunWithContextBytes` -> `RunBytes`

Do most people like to rename it now? And I think if we do so, we should also rename `RunContext`, it's not a context. What's the proper name?

**Comment by wxiaoguang** [comment]
And the name `Run` is too short. I prefer to use `RunCmd`, `RunStdString`, `RunStdBytes`, and use `CmdParams` instead of `RunContext`

The longer names are friendly to be grep-ed out during refactoring.

**Comment by lunny** [comment]
> And the name `Run` is too short. I prefer to use `RunCmd`, `RunStdString`, `RunStdBytes`, and use `CmdParams` instead of `RunContext`
> 
> The longer names are friendly to be grep-ed out during refactoring.

I don't think `command.RunCmd` is a good name.

**Comment by wxiaoguang** [comment]
Hmmm, `RunCmd` is not good indeed. Do we have choices for a two-word name?

**Comment by wxiaoguang** [comment]
I can not get a better name now. So how about:

* `cmd.Run(&CmdOpts{Dir: "..."})`
* `cmd.RunStdString(&CmdOpts{Dir: "..."})`   // std means it handles stdout/stderr
* `cmd.RunStdBytes(&CmdOpts{Dir: "..."})`

**Comment by lunny** [comment]
> I can not get a better name now. So how about:
> 
> * `cmd.Run(&git.CmdOpts{Dir: "..."})`
> * `cmd.RunStdString(&git.CmdOpts{Dir: "..."})`   // std means it handles stdout/stderr
> * `cmd.RunStdBytes(&git.CmdOpts{Dir: "..."})`
> 
> If no objection, I will do it.

I still think `Std` will not help for understanding the method meaning. And the options is for `Run` not for `Command`. `Command` have many `Args` methods. So `CmdOpts` is not better than `RunOpts`. In fact, I think move command to `git/command` sub package is better, so that `RunOpts` will be more clear than `CmdOpts`.

**Comment by wxiaoguang** [comment]
I pushed a demo for:

* `cmd.Run(&git.RunOpts{Dir: "..."})`
* `cmd.RunStdString(&git.RunOpts{Dir: "..."})`   // std means it handles stdout/stderr
* `cmd.RunStdBytes(&git.RunOpts{Dir: "..."})`

I think the `Std` is a must, because these two functions are totally different from `Run`, they handle stdin/stdout internally, and the returned error is totally different.

**Comment by wxiaoguang** [comment]
@lunny a real example, when you read:

```go
err := cmd.Run(...)
_, _, err := cmd.RunString(...)

// then
if strings.Contains(err.Error(), "fatal: ...")) {
}
```

Do you know that these 2 errors are different? The second one contains `stderr` output.

So `_, _, err := cmd.RunStdString(...)` is more clear in such case.

**Review by lunny — APPROVED** [review]

**Pull request merged**